_posts/2022-08-05-hacking-for-cruise-control-ui.md

```diff
@@ -0,0 +1,334 @@
+
+RUN curl -LO https://github.com/linkedin/cruise-control-ui/releases/download/v${CC_UI_VERSION}/cruise-control-ui-${CC_UI_VERSION}.tar.gz; \
+    tar xvfz cruise-control-ui-${CC_UI_VERSION}.tar.gz -C ${CC_UI_HOME} --strip-components=2; \
+    rm -f cruise-control-ui-${CC_UI_VERSION}.tar.gz*; \
+    echo "dev,dev,/kafkacruisecontrol/" > "${CC_UI_HOME}"static/config.csv;
+
+USER 1001
+```
+
+Then building and pushing the image to a container registry:
+```
+# When building, tag the image with the name of a container registry
+# that is accessible by the Kubernetes cluster.
+docker build . -t <registry>/<cruise-control-with-ui-image-name>
+
+# Push the image to that container registry
+docker push <registry>/<cruise-control-with-ui-image-name>
+```
+
+### Deploying the custom Cruise Control pod
+
+We can deploy our custom Cruise Control pod in one of two ways:
+
+* Deployment via `Kafka` custom resource
+* Deployment via Strimzi Cluster Operator `Deployment`
+
+#### Deployment via `Kafka` custom resource
+
+This method will restart the Cruise Control instance associated with our `Kafka` resource.
+We can edit our `Kafka` resource like this:
+
+```
+kubectl edit kafka <kafka-custom-resource-name>
+```
+
+Updating the `cruiseControl.image` section of the spec with the custom Cruise Control image:
+
+```
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+   ...
+spec:
+   cruiseControl:
+     image: <registry>/<cruise-control-with-ui-image-name>
+   ...
+```
+
+#### Deployment via Strimzi Cluster Operator `Deployment`
+
+This method will restart the Cluster Operator pod and all Cruise Control instances.
+We can edit our Strimzi Cluster Operator `Deployment` like this:
+
+```
+kubectl edit deployment strizi-cluster-operator
+```
+
+Updating the env section of the spec with the custom Cruise Control image:
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: strimzi-cluster-operator
+  ...
+spec:
+  ...
+  template:
+    ...
+    spec:
+      containers:
+      - args:
+        - /opt/strimzi/bin/cluster_operator_run.sh
+        env:
+          ...
+        - name: STRIMZI_DEFAULT_CRUISE_CONTROL_IMAGE
+          value: <registry>/<cruise-control-with-ui-image-name>
+
+```
+
+After updating the deployment, the Cluster Operator will pull the new custom Cruise Control image and run it.
+
+## METHOD TWO: Creating a dedicated Cruise Control UI pod
+
+With our Strimzi-managed Kafka cluster deployed, we can now create a Cruise Control UI pod to run alongside it.
+
+![](/assets/images/posts/2022-10-31-hacking-for-cruise-control-ui-2.png)
+
+### Inside the Cruise Control UI container
+
+Following the instructions on the [Cruise Control UI wiki](https://github.com/linkedin/cruise-control-ui/wiki/CCFE-(Dev-Mode)---Docker), we can configure an Nginx server as a reverse-proxy to forward Cruise Control UI requests to the Cruise Control REST API.
+
+![](/assets/images/posts/2022-10-31-hacking-for-cruise-control-ui-4.png)
```

Comment:
The picture presents the UI as a separate process communicating with the Nginx which proxies the requests to the Cruise Control API. Is that the case? Or is the UI just a bunch of static files served by Nginx? That seems to be suggested by the Dockerfile.

Reply:
The UI is just static files served over NGNIX. Updated the diagram accordingly, let me know what you think!